app/src/main/java/de/westnordost/streetcomplete/screens/main/RecentLocationStore.kt

```diff
@@ -0,0 +1,49 @@
+package de.westnordost.streetcomplete.screens.main
+
+import android.location.Location
+import de.westnordost.streetcomplete.util.math.EARTH_RADIUS
+import kotlin.math.PI
+import kotlin.math.abs
+import kotlin.math.cos
+
+class RecentLocationStore {
+    private val recentLocations: MutableList<Location> = mutableListOf()
+
+    /** returns a sequence of recent locations, with some minimum time and distance from each other */
+    fun get(): Sequence<Location> = synchronized(recentLocations) {
+        var previousLocation: Location? = null
+        recentLocations.asReversed().asSequence().filter {
```

Comment:
What is synchronized is the creation of the sequence, but the sequence is based on a non-copied version of `recentLocations` which means that `recentLocations` can be iterated through outside of the synchronized block. So, the synchronized block does not actually protect against race conditions. In fact, I believe this can lead to concurrent modification exceptions when the list is iterated while new locations are added.

I understand that you do not want to iterate the whole list and calculate `isTooCloseTo` on it, but this approach won't work for the above reason.
Couldn't you do this check in `add` instead?  (Should be faster, too)

Reply:
The sequence should always start with the most recent location, which is not guaranteed when checking in `add`. In worst case the user could be close to the location, but still get the popup.

Reply:
That doesn't seem to be a big problem. Just increase the maximum survey distance constant by a bit.